### DISCUSSION TITLE
Submit a workflow and return outputs with a single API request

### DISCUSSION THREAD

**Discussion by conorsheehan-ow** [argoproj/argo-workflows#12995] [Q&A] [answered]
I have a workflow that returns some data. It's very fast so should complete before any http request timeout.

I'd like to trigger the workflow and get the response in a single API request. 
Looks like this is possible via the CLI: https://argo-workflows.readthedocs.io/en/latest/cli/argo_wait/
But I can't seem to find a way to do this over the API?

Is it possible to submit a workflow and get the response in a single API request? 
Ideally in javascript using axios. e.g. code I have to submit a workflow looks like this:

```Javascript
const namespace = process.env.ARGO_NAMESPACE;
const templateName = "very-fast-workflow";
const workflow = {
  apiVersion: 'argoproj.io/v1alpha1',
  kind: 'Workflow',
  spec: {
    arguments: {
      parameters: [
        {
          name: 'json',
          value: JSON.stringify(jsonArgs),
        },
      ],
    },
    workflowTemplateRef: {
      name: templateName,
    },
  },
};
const payload = { namespace, workflow };
// this resolves once the workflow is submitted, but I want to wait until it finishes and return the results
await axiosArgo.post(`${process.env.ARGO_URL}/api/v1/workflows/${namesapce}`, payload);
```

**Comment by Joibel** [comment]
The CLI would make multiple requests to achieve your goal. Essentially, `submit`, `wait`, then `get results`.
You haven't given a reason why you require all of this in a single API call.

**Reply by conorsheehan-ow** [reply]
I'm trying to migrate some python analytics from an old internal tool built on top of django to argo workflows.
We have two concepts there: 1. jobs and 2. functions.
1. Jobs are fine, I've already moved them to argo workflows and everything works, because we expect them to be long running / async.
2. Functions have a problem, because the old implementation expects them to be short / synchronous, returning the output as part of the http response.

I'm trying to avoid using django/flask/similar in favour of argo workflows, but looks like it just doesn't fit the usecase.
I was hoping to get similar behaviour for the functions with argo but I guess it's not possible or designed to work that way.

Thanks all for the help though!

**Comment by agilgur5** [comment, marked as answer]
> I'd like to trigger the workflow and get the response in a single API request.

You can't. Workflows are asynchronous. In general, most k8s resources are. For instance, when you create a Pod, it does not get instantly run, it only gets scheduled to run. Workflows are the same.

> Looks like this is possible via the CLI: https://argo-workflows.readthedocs.io/en/latest/cli/argo_wait/

`argo wait` [performs](https://github.com/argoproj/argo-workflows/blob/ad823a51fabbe29966e27884e72a5dc21c8b9ada/cmd/argo/commands/common/wait.go#L43) a k8s `watch`. So an `argo submit` followed by an `argo wait` is 2 API requests.

> But I can't seem to find a way to do this over the API?

You can add [`listOptions.watch`](https://github.com/argoproj/argo-workflows/blob/ad823a51fabbe29966e27884e72a5dc21c8b9ada/api/openapi-spec/swagger.json#L2840) to a get request to watch a resource via the API.
